asterius/src/Asterius/Types/SymbolMap.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module      :  Asterius.Types.SymbolMap
+-- Copyright   :  (c) 2018 EURL Tweag
+-- License     :  All rights reserved (see LICENCE file in the distribution).
+--
+-- The @'SymbolMap' a@ type represents a finite map/dictionary from keys of
+-- type 'EntitySymbol' to values of type @a@. Internally it is represented as
+-- an 'IM.IntMap' of , where the key of the unique of the 'EntitySymbol' is used
+-- for the indexing. In order to be able to access the 'EntitySymbol'
+-- corresponding to each entry, we also store it alongside each element.
+module Asterius.Types.SymbolMap
+  ( -- * SymbolMap type
+    SymbolMap,
+
+    -- * Construction
+    empty,
+    singleton,
+
+    -- * Query
+    member,
+    lookup,
+    (!),
+
+    -- * Size
+    size,
+
+    -- * Insertion
+    insert,
+
+    -- * Filtering
+    filterWithKey,
+    filter,
+    restrictKeys,
+
+    -- * Folds and Maps
+    foldrWithKey',
+    mapWithKey,
+    mapAccum,
+    mapKeys,
+
+    -- * Conversion
+    elems,
+    keys,
+    keysSet,
+
+    -- ** Lists
+    toList,
+    fromList,
+
+    -- ** Maps
+    toMap,
+    fromMap,
+  )
+where
+
+import Asterius.Binary.Orphans ()
+import Asterius.Types.EntitySymbol
+import Binary
+import Control.Monad
+import Data.Data
+import qualified Data.IntMap as IM
+import qualified Data.IntSet as IS
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import GHC.Exts (IsList (..))
+import GHC.Stack
+import Unique
+import Prelude hiding (filter, lookup)
+
+-- | A map from 'EntitySymbol's to values @a@.
+newtype SymbolMap a = SymbolMap (IM.IntMap (EntitySymbol, a))
+  deriving newtype (Eq)
+  deriving stock (Data)
+
+instance Show a => Show (SymbolMap a) where
+  showsPrec d m =
+    showParen (d > 10) $
+      showString "fromList " . shows (toListSM m)
+
+instance Semigroup (SymbolMap a) where
+  SymbolMap m1 <> SymbolMap m2 = SymbolMap $ m1 <> m2
+
+instance Monoid (SymbolMap a) where
+  mempty = empty
+  mappend = (<>)
+
+instance Binary a => Binary (SymbolMap a) where
+  -- We are lazy on both at the moment, hence
+  -- the usage of lazyPut for both (instead of put_).
+
+  put_ bh m =
+    put_ bh (size m)
+      *> forM_ (toListSM m) (\(k, v) -> lazyPut bh k *> lazyPut bh v)
+
+  -- We are lazy on both at the moment, hence
+  -- the usage of lazyGet for both (instead of get).
+
+  get bh = fromListSM <$> do
+    n <- get bh
+    replicateM n $ (,) <$> lazyGet bh <*> lazyGet bh
+
+instance IsList (SymbolMap a) where
+  type Item (SymbolMap a) = (EntitySymbol, a)
+  fromList = fromListSM
+  toList = toListSM
+
+-- ----------------------------------------------------------------------------
+
+-- | /O(1)/. The empty map.
+{-# INLINE empty #-}
+empty :: SymbolMap a
+empty = SymbolMap IM.empty
+
+-- | /O(1)/. A map containing a single association.
+{-# INLINE singleton #-}
+singleton :: EntitySymbol -> a -> SymbolMap a
+singleton k e = SymbolMap $ IM.singleton (getKey $ getUnique k) (k, e)
```

Comment:
There're multiple `getKey $ getUnique k` usages in this module, better factor it out to an inline function.

Reply:
Done (I created `getKeyES`).